packages/eslint-plugin/docs/rules/typedef.md

```diff
@@ -0,0 +1,364 @@
+    ]
+  }
+}
+```
+
+- Type annotations on arrow function parameters are not required
+- Type annotations on variables are required
+- Options otherwise adhere to the defaults
+
+### arrayDestructuring
+
+Whether to enforce type annotations on variables declared using array destructuring.
+
+Examples of **incorrect** code with `{ "arrayDestructuring": true }`:
+
+```ts
+const [a] = [1];
+const [b, c] = [1, 2];
+```
+
+Examples of **correct** code with `{ "arrayDestructuring": true }`:
+
+```ts
+const [a]: number[] = [1];
+const [b]: [number] = [2];
+const [c, d]: [boolean, string] = [true, 'text'];
+```
+
+### arrowCallSignature
+
+Whether to enforce type annotations for return types of arrow functions.
+
+Examples of **incorrect** code with `{ "arrowCallSignature": true }`:
+
+```ts
+const returnsVoid = () => {};
+const returnsString = () => 'text';
+
+['hello', 'world'].map(text => text.length);
+
+const mapper = {
+  map: text => text + '...',
+};
+```
+
+Examples of **correct** code with `{ "arrowCallSignature": true }`:
+
+```ts
+const returnsVoid = (): void => {};
+const returnsString = (): string => 'text';
+
+['hello', 'world'].map((text): number => text.length);
+
+const mapper = {
+  map: (text): string => text + '...',
+};
+```
+
+### arrowParameter
+
+Whether to enforce type annotations for parameters of arrow functions.
+
+Examples of **incorrect** code with `{ "arrowParameter": true }`:
+
+```ts
+const logsSize = size => console.log(size);
+
+['hello', 'world'].map(text => text.length);
+
+const mapper = {
+  map: text => text + '...',
+};
+```
+
+Examples of **correct** code with `{ "arrowParameter": true }`:
+
+```ts
+const logsSize = (size: number) => console.log(text);
+
+['hello', 'world'].map((text: string) => text.length);
+
+const mapper = {
+  map: (text: string) => text + '...',
+};
+```
+
+### callSignature
+
+Whether to enforce type annotations for return types of functions and methods.
+
+Examples of **incorrect** code with `{ "callSignature": true }`:
+
+```ts
+function logsSize(size: number) {
+  console.log(size);
+}
+
+const doublesSize = function(size: number) {
+  return size * 2;
+};
+
+const divider = {
+  dividesSize(size: number) {
+    return size / 2;
+  },
+  doesNothing: function() {
+    console.log('...');
+  },
+};
+
+class Logger {
+  log(text: string) {
+    console.log('>', text);
+    return true;
+  }
+}
+```
+
+Examples of **correct** code with `{ "callSignature": true }`:
+
+```ts
+function logsSize(size: number): void {
+  console.log(size);
+}
+
+const doublesSize = function(size: number): number {
+  return size * 2;
+};
+
+const divider = {
+  dividesSize(size: number): number {
+    return size / 2;
+  },
+  doesNothing: function(): void {
+    console.log('...');
+  },
+};
+
+class Logger {
+  log(text: string): boolean {
+    console.log('>', text);
+    return true;
+  }
+}
+```
+
+### memberVariableDeclaration
+
+Whether to enforce type annotations on member variables of classes.
```

Comment:
I was REALLY surprised to learn that typescript doesn't enforce this with `noImplicitAny`.

```TS
class Foo {
    prop; // DOES NOT ERROR!!!
}

const x = new Foo()
x.prop; // typeof === any
```